compiler/src/dotty/tools/dotc/transform/Mixin.scala

```diff
@@ -121,60 +128,70 @@ class Mixin extends MiniPhase with SymTransformer { thisPhase =>
       sym.copySymDenotation(
         name = nme.TRAIT_CONSTRUCTOR,
         info = MethodType(Nil, sym.info.resultType))
+    else if sym.is(Trait, butNot = JavaDefined) then
+      val classInfo = sym.asClass.classInfo
+      val decls1 = classInfo.decls.cloneScope
+      var modified: Boolean = false
+      for (decl <- classInfo.decls)
+        // !decl.isClass avoids forcing nested traits, preventing cycles
+        if !decl.isClass && needsTraitSetter(decl) then
+          val setter = makeTraitSetter(decl.asTerm)
+          decls1.enter(setter)
+          modified = true
+      if modified then
+        sym.copySymDenotation(
+          info = classInfo.derivedClassInfo(decls = decls1))
+      else
+        sym
     else
       sym
+  end transformSym
 
-  private def initializer(sym: Symbol)(using Context): TermSymbol = {
-    if (sym.is(Lazy)) sym
-    else {
-      val initName = InitializerName(sym.name.asTermName)
-      sym.owner.info.decl(initName).symbol
-        .orElse(
-          ctx.newSymbol(
-            sym.owner,
-            initName,
-            Protected | Synthetic | Method,
-            sym.info,
-            coord = sym.coord).enteredAfter(thisPhase))
-    }
-  }.asTerm
+  private def wasOneOf(sym: Symbol, flags: FlagSet)(using Context): Boolean =
+    atPhase(thisPhase) { sym.isOneOf(flags) }
+
+  private def needsTraitSetter(sym: Symbol)(using Context): Boolean =
+    sym.isGetter && !wasOneOf(sym, DeferredOrLazy | ParamAccessor)
+      && atPhase(thisPhase) { !sym.setter.exists }
+      && !sym.info.resultType.isInstanceOf[ConstantType]
+
+  private def makeTraitSetter(getter: TermSymbol)(using Context): Symbol =
+    getter.copy(
+      name = Mixin.traitSetterName(getter),
+      flags = Method | Accessor | Deferred,
+      info = MethodType(getter.info.resultType :: Nil, defn.UnitType))
 
   override def transformTemplate(impl: Template)(using Context): Template = {
     val cls = impl.symbol.owner.asClass
     val ops = new MixinOps(cls, thisPhase)
     import ops._
 
     def traitDefs(stats: List[Tree]): List[Tree] = {
-      val initBuf = new mutable.ListBuffer[Tree]
-      stats.flatMap({
-        case stat: DefDef if stat.symbol.isGetter && !stat.rhs.isEmpty && !stat.symbol.is(Lazy) =>
-          // make initializer that has all effects of previous getter,
-          // replace getter rhs with empty tree.
-          val vsym = stat.symbol
-          val isym = initializer(vsym)
-          val rhs = Block(
-            initBuf.toList.map(_.changeOwnerAfter(impl.symbol, isym, thisPhase)),
-            stat.rhs.changeOwnerAfter(vsym, isym, thisPhase).wildcardToDefault)
-          initBuf.clear()
-          cpy.DefDef(stat)(rhs = EmptyTree) :: DefDef(isym, rhs) :: Nil
+      stats.flatMap {
+        case stat: DefDef if needsTraitSetter(stat.symbol) =>
+          // add a trait setter for this getter
+          stat :: DefDef(stat.symbol.traitSetter.asTerm, EmptyTree) :: Nil
```

Comment:
I'm not sure if it would fit here, but an alternative way to go from one def to two defs is to use `transformDefDef` and return a `Thicket(defA, defB)`, that would be more efficient that using flatMap when most cases have to construct a singleton list

Reply:
Since `traitDefs` is part of the larger logic of `transformTemplate`, it seems that this suggestion is not so easy to apply. I would rather keep the existing structure, especially since it comes from before this PR.